# Jupyter - Day 6 - Section 002
## Lecture 6 - Multiple Linear Regression

In the last few lectures, we have focused on single input linear regression, that is, fitting models of the form 

$$
Y =  \beta_0 +  \beta_1 X + \varepsilon
$$

In this lab, we will continue to use two different tools for linear regression. 
- [Scikit learn](https://scikit-learn.org/stable/index.html) is arguably the most used tool for machine learning in python 
- [Statsmodels](https://www.statsmodels.org) provides many of the statisitcial tests we've been learning in class

In [2]:
# As always, we start with our favorite standard imports. 

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
%matplotlib inline
import seaborn as sns
import statsmodels.formula.api as smf
from sklearn.linear_model import LinearRegression

## Multiple linear regression 
Next we get some code up and running that can do linear regression with multiple input variables, that is when the model is of the form

$$
Y =  \beta_0 +  \beta_1 X_1 +  \beta_2 X_2 + \cdots +  \beta_pX_p + \varepsilon
$$

In [3]:
from sklearn.datasets import load_diabetes
diabetes = load_diabetes(as_frame=True)
diabetes_df = pd.DataFrame(diabetes.data, columns = diabetes.feature_names)
diabetes_df['target'] = pd.Series(diabetes.target)

diabetes_df

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6,target
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646,151.0
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204,75.0
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930,141.0
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362,206.0
4,0.005383,-0.044642,-0.036385,0.021872,0.003935,0.015596,0.008142,-0.002592,-0.031988,-0.046641,135.0
...,...,...,...,...,...,...,...,...,...,...,...
437,0.041708,0.050680,0.019662,0.059744,-0.005697,-0.002566,-0.028674,-0.002592,0.031193,0.007207,178.0
438,-0.005515,0.050680,-0.015906,-0.067642,0.049341,0.079165,-0.028674,0.034309,-0.018114,0.044485,104.0
439,0.041708,0.050680,-0.015906,0.017293,-0.037344,-0.013840,-0.024993,-0.011080,-0.046883,0.015491,132.0
440,-0.045472,-0.044642,0.039062,0.001215,0.016318,0.015283,-0.028674,0.026560,0.044529,-0.025930,220.0


We first model `target = beta_0 + beta_1 *s1 + beta_2 * s5` using `scikitlearn`.

In [4]:
X = diabetes_df[['s1','s5']].values
y = diabetes_df['target'].values

multireg = LinearRegression() #<----- notice I'm using exactly the same command as above
multireg.fit(X,y)

print(multireg.coef_)
print(multireg.intercept_)

[-175.71107989 1006.71695008]
152.13348416289585


&#9989; **<font color=red>Q:</font>** What are the values for $\beta_0$, $\beta_1$, and $\beta_2$? Write an interpretation for the $\beta_2$ value in this data set. 

Your answer here

We next model `target = beta_0 + beta_1 *s1 + beta_2 * s5` using `statsmodels`. Do you get the same model?

In [5]:
# multiple least squares with statsmodel
multiple_est = smf.ols('target ~ s1 + s5', diabetes_df).fit()
multiple_est.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 target   R-squared:                       0.329
Model:                            OLS   Adj. R-squared:                  0.326
Method:                 Least Squares   F-statistic:                     107.6
Date:                Mon, 08 Sep 2025   Prob (F-statistic):           9.63e-39
Time:                        10:49:16   Log-Likelihood:                -2459.0
No. Observations:                 442   AIC:                             4924.
Df Residuals:                     439   BIC:                             4936.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    152.1335      3.011     50.528      0.000     146.216     158.051
s1          -175.7111     73.872     -2.379      0.018    -320.898     -30.524
s5          1006.7170     73.872     13.628      0.000     861.530    1151.904
==============================================================================
Omnibus:                       10.295   Durbin-Watson:                   2.025
Prob(Omnibus):                  0.006   Jarque-Bera (JB):               10.497
Skew:                           0.356   Prob(JB):                      0.00526
Kurtosis:                       2.748   Cond. No.                         30.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

&#9989; **<font color=red>Q:</font>** What is the predicted model? How much trust can we place in the estimates?

*Your answer here*

##ANSWER## 

$y = 152.133- 175.71 * s1 + 1006*s5$

All low p-values, so high confidence.

&#9989; **<font color=red>Q:</font>** Run the linear regression to predict `target` using all the other variables. What do you notice about the different terms? Are some more related than others? 

*Your answer here*

In [6]:
##ANSWER## 

cols = list(diabetes_df.columns.copy())
cols = cols[:-1]
print(cols)


# Linear regression with Scikit-learn
multiple_X = diabetes_df[cols]
multiple_y = diabetes_df[['target']]

regressor = LinearRegression()
regressor.fit(multiple_X, multiple_y)
print(regressor.coef_)
print(regressor.intercept_)

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']
[[ -10.0098663  -239.81564367  519.84592005  324.3846455  -792.17563855
   476.73902101  101.04326794  177.06323767  751.27369956   67.62669218]]
[152.13348416]


In [7]:
##ANSWER## 
# Using statsmodels
model = 'target ~ ' + ' + '.join(cols)
print(model)

multiple_est = smf.ols(model, diabetes_df).fit()
multiple_est.summary()

target ~ age + sex + bmi + bp + s1 + s2 + s3 + s4 + s5 + s6


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 target   R-squared:                       0.518
Model:                            OLS   Adj. R-squared:                  0.507
Method:                 Least Squares   F-statistic:                     46.27
Date:                Mon, 08 Sep 2025   Prob (F-statistic):           3.83e-62
Time:                        10:50:55   Log-Likelihood:                -2386.0
No. Observations:                 442   AIC:                             4794.
Df Residuals:                     431   BIC:                             4839.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    152.1335      2.576     59.061      0.000     147.071     157.196
age          -10.0099     59.749     -0.168      0.867    -127.446     107.426
sex         -239.8156     61.222     -3.917      0.000    -360.147    -119.484
bmi          519.8459     66.533      7.813      0.000     389.076     650.616
bp           324.3846     65.422      4.958      0.000     195.799     452.970
s1          -792.1756    416.680     -1.901      0.058   -1611.153      26.802
s2           476.7390    339.030      1.406      0.160    -189.620    1143.098
s3           101.0433    212.531      0.475      0.635    -316.684     518.770
s4           177.0632    161.476      1.097      0.273    -140.315     494.441
s5           751.2737    171.900      4.370      0.000     413.407    1089.140
s6            67.6267     65.984      1.025      0.306     -62.064     197.318
==============================================================================
Omnibus:                        1.506   Durbin-Watson:                   2.029
Prob(Omnibus):                  0.471   Jarque-Bera (JB):                1.404
Skew:                           0.017   Prob(JB):                        0.496
Kurtosis:                       2.726   Cond. No.                         227.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

&#9989; **<font color=red>Q:</font>** Earlier you determined the p-value for the `s1` variable when we only used `s1` to predict `target`. What changed about the p-value for `s1` now where it is part of a regression using all the variables. Why?

In [ ]:
# Your answer here

In [8]:
##ANSWER## 
just = smf.ols('target ~ s1', diabetes_df).fit()
just.summary().tables[1]

,coef,std err,t,P>|t|,[0.025,0.975]
Intercept,152.1335,3.588,42.405,0.000,145.082,159.185
s1,343.2545,75.426,4.551,0.000,195.015,491.494


In [13]:
##ANSWER##
# Note that entries 4 and 5 are highly correlated with each other, 
# so 's1' and 's2' are compensating for each other
# im = plt.matshow(diabetes_df.corr())
# plt.colorbar(im)

sns.heatmap(diabetes_df.corr(), annot=True, cmap='coolwarm')
# plt.show()

<Axes: >

In [11]:
##ANSWER##
diabetes_df.corr()

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6,target
age,1.000000,0.173737,0.185085,0.335428,0.260061,0.219243,-0.075181,0.203841,0.270774,0.301731,0.187889
sex,0.173737,1.000000,0.088161,0.241010,0.035277,0.142637,-0.379090,0.332115,0.149916,0.208133,0.043062
bmi,0.185085,0.088161,1.000000,0.395411,0.249777,0.261170,-0.366811,0.413807,0.446157,0.388680,0.586450
bp,0.335428,0.241010,0.395411,1.000000,0.242464,0.185548,-0.178762,0.257650,0.393480,0.390430,0.441482
s1,0.260061,0.035277,0.249777,0.242464,1.000000,0.896663,0.051519,0.542207,0.515503,0.325717,0.212022
s2,0.219243,0.142637,0.261170,0.185548,0.896663,1.000000,-0.196455,0.659817,0.318357,0.290600,0.174054
s3,-0.075181,-0.379090,-0.366811,-0.178762,0.051519,-0.196455,1.000000,-0.738493,-0.398577,-0.273697,-0.394789
s4,0.203841,0.332115,0.413807,0.257650,0.542207,0.659817,-0.738493,1.000000,0.617859,0.417212,0.430453
s5,0.270774,0.149916,0.446157,0.393480,0.515503,0.318357,-0.398577,0.617859,1.000000,0.464669,0.565883
s6,0.301731,0.208133,0.388680,0.390430,0.325717,0.290600,-0.273697,0.417212,0.464669,1.000000,0.382483


![Stop Icon](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1e/Vienna_Convention_road_sign_B2a.svg/180px-Vienna_Convention_road_sign_B2a.svg.png)

Great, you got to here! Hang out for a bit, there's more lecture before we go on to the next portion. 

In [14]:
# We're going to use the (fake) data set used in the beginning of the book. 
# you may need to find the data on the course website and change the path below!
advertising_df = pd.read_csv('../../../DataSets/Advertising.csv', index_col = 0)
advertising_df

,TV,Radio,Newspaper,Sales
1,230.1,37.8,69.2,22.1
2,44.5,39.3,45.1,10.4
3,17.2,45.9,69.3,9.3
4,151.5,41.3,58.5,18.5
5,180.8,10.8,58.4,12.9
...,...,...,...,...
196,38.2,3.7,13.8,7.6
197,94.2,4.9,8.1,9.7
198,177.0,9.3,6.4,12.8
199,283.6,42.0,66.2,25.5


# Q1: Hypothesis Test

&#9989; **<font color=red>Do this:</font>** Use the `statsmodels` package to fit the model 
$$
\texttt{Sales} = \beta_0 + \beta_1 \cdot \texttt{TV} + \beta_2\cdot \texttt{Radio} + \beta_3\cdot \texttt{Newspaper}
$$
What is the equation for the model learned? 

In [14]:
# Your code here

In [15]:
##ANSWER##
est = smf.ols('Sales ~ TV + Radio + Newspaper', advertising_df).fit()
est.params

Intercept    2.938889
TV           0.045765
Radio        0.188530
Newspaper   -0.001037
dtype: float64

In [16]:
##ANSWER##

modelstring = 'Sales = '
modelstring += str(round(est.params[0],2)) 
modelstring += ' + ' + str(round(est.params[1],2)) + ' * TV' 
modelstring += ' + ' + str(round(est.params[2],2)) + ' * Radio' 
modelstring += ' + ' + str(round(est.params[3],2)) + ' * Newspaper' 

modelstring

C:\Users\mengs\AppData\Local\Temp\ipykernel_29768\3674854890.py:4: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  modelstring += str(round(est.params[0],2))
C:\Users\mengs\AppData\Local\Temp\ipykernel_29768\3674854890.py:5: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  modelstring += ' + ' + str(round(est.params[1],2)) + ' * TV'
C:\Users\mengs\AppData\Local\Temp\ipykernel_29768\3674854890.py:6: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[p

'Sales = 2.94 + 0.05 * TV + 0.19 * Radio + -0.0 * Newspaper'

&#9989; **<font color=red>Do this:</font>** Use the `summary` command for the trained model class to determine the F-statistic for this model. 
- What are the null and alternative hypotheses for the test this statistic is used for? 
- What is your conclusion of the test given this F-score?

In [17]:
##ANSWER##
est.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Mon, 08 Sep 2025   Prob (F-statistic):           1.58e-96
Time:                        10:55:25   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
Radio          0.1885      0.009     21.893      0.000       0.172       0.206
Newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [18]:
##ANSWER##
# If we just want the p-value
est.f_pvalue
# Note very small, so we will reject the null hypothesis and conclude that SOME 
# input variable is related to the output variable. Note that this does not tell
# us which one it is

1.575227256092437e-96

![Stop Icon](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1e/Vienna_Convention_road_sign_B2a.svg/180px-Vienna_Convention_road_sign_B2a.svg.png)

Great, you got to here! Hang out for a bit, there's more lecture before we go on to the next portion. 

# Q2: Subsets of variables

&#9989; **<font color=red>Q:</font>** List all 6 subsets of the three variables being used. 

*Your answer here*

##ANSWER##
- TV
- Radio
- Newspaper
- TV, Radio
- TV, Newspaper
- Radio, Newspaper 
- TV, Radio, Newspaper

&#9989; **<font color=red>Do this:</font>** Below is a command to get the RSS for the statsmodel linear fit. For each of the subsets listed above, what is the RSS for the learned model? Which is smallest? 

In [21]:
est.ssr

556.8252629021872

In [22]:
##ANSWER##
import itertools 

variables = ['TV', 'Radio', 'Newspaper']
for i in range(1,4): # Want 1, 2, 3
    for subset in itertools.combinations(variables,i):
        model = 'Sales ~ ' + " + ".join(subset)
        print(model)
        est = smf.ols(model, advertising_df).fit()
        print(np.round(est.ssr,2))
    
# Best job done by all three or TV and Radio

Sales ~ TV
2102.53
Sales ~ Radio
3618.48
Sales ~ Newspaper
5134.8
Sales ~ TV + Radio
556.91
Sales ~ TV + Newspaper
1918.56
Sales ~ Radio + Newspaper
3614.84
Sales ~ TV + Radio + Newspaper
556.83




-----
### Congratulations, we're done!

Initially Written by Dr. Liz Munch, Adapted by Dr. Mengsen Zhang, Michigan State University
<a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 International License</a>.

In [23]:
##ANSWER##
#This cell runs the converter which removes ANSWER fields, renames the notebook and cleans out output fields. 
from jupyterinstruct import InstructorNotebook
import os
this_notebook = os.path.basename(globals()['__vsc_ipynb_file__'])

studentnotebook = InstructorNotebook.makestudent(this_notebook)

InstructorNotebook.validate(studentnotebook)


<IPython.core.display.Javascript object>

Myfilename CMSE381-002-Lec06-MultLinReg-INSTRUCTOR.ipynb


CMSE381-002_Lec06_MultLinReg.ipynb


Validating Notebook ./CMSE381-002_Lec06_MultLinReg.ipynb

   ERROR: Image LINK not found - http://creativecommons.org/licenses/by-nc/4.0/
   ERROR: Image LINK not found - http://creativecommons.org/licenses/by-nc/4.0/


3